### ISSUE TITLE
AWS RDS Failover hangs hangfire for 3 days and it cannot start again 10K+ pending jobs

### ISSUE THREAD

**Issue by sseyalioglu** [HangfireIO/Hangfire#1849]
Hi,

I recently upgraded my Aurora DB (Mysql) to the latest version and during the process DB rebooted, failed over etc. But one thing I was unable notice was, hangfire process get into partial working state on all servers (4)
It was able to queue jobs but not able to process anything at all. 
As a result, I restarted the instances so it to be kicked in. 
Here are the problems happened as far as I understand

1. All servers sent so many requests to DB which made DB get into 100% and not stopping to run the commands with a good amount of queries being in somewhat dead-lock state. 
2. Causing DB to be 100% CPU.
3. Followed with timeouts and exceptions related as it can be seen here :
![image](https://user-images.githubusercontent.com/2631639/115344686-94306a80-a162-11eb-97a7-c62b5b1ad31d.png)
4. I suspect 1 & 2 start the issue in general, causing so many bulk hits to DB
5. Due to so many process getting stuck in DB, timeouts happen just like in "e"

To handle the case, I made another database so systems will use fresh empty DB and it worked fine. But problem is I want to process locked up 10K+ jobs so I put one of the servers back to old DB connection. But now, this single server gets locked up and stuck just like the screenshot above. Problem is that, I think it uses so much memory that I cannot even get connected to the server via SSH.  the only option I have is to reboot the server and  once it is up, stop the service so I can modify connectionstring so server does respond and process properly. 

Looking for the following

1. Trying to prevent such condition for anyone in the future. 
2. A way to process my 10+ jobs
3. A good suggestion on alerts for hangfire so I will know that it is not working.

For #1,  DB failover shall not stop processing completely (like defined above, taking new queue but not processing any) see below
![image](https://user-images.githubusercontent.com/2631639/115345365-8c24fa80-a163-11eb-95e1-f55a3fef8382.png)

Again for #1: Once DB is available again, do not bombard with whatever is having to completely lock up the systems and be in not able to process any DB operations like can be seen below:
![image](https://user-images.githubusercontent.com/2631639/115345598-e2923900-a163-11eb-9f9f-4ab62fed3c9c.png)

**Comment by sseyalioglu** [comment]
@odinserj any inputs here? May be I can run something to clean whatever is getting stuck directly in DB so hangfire can process the rest properly.

**Comment by odinserj** [comment]
This is very likely to be an issue with the community-driven storage implementation for MySQL. Please report the issue in the corresponding repository – job storage implementation is the central piece of background processing, and everything depends on the absence of bugs in this component.

**Comment by sseyalioglu** [comment]
Thanks, will do.

**Comment by sseyalioglu** [comment]
Posted there [https://github.com/arnoldasgudas/Hangfire.MySqlStorage/issues/81](https://github.com/arnoldasgudas/Hangfire.MySqlStorage/issues/81) and will update here if resolved.

**Comment by sseyalioglu** [comment]
@odinserj, I recently see system not picking up jobs too after this upgrade. Do you think, that is also related to database connector? System shows queued jobs but seems like something wrong with "Hangfire.Processing.BackgroundExecution"
Please see that I have no logs for long time on this, which doesn't look normal to me.
![image](https://user-images.githubusercontent.com/2631639/115452844-ccbc5c80-a1d3-11eb-9bc5-12aff493734b.png)

Addiitionally server becomes unresponsive in dashboard and SSH although from logs I see that it process work (not related to hangfire)
![image](https://user-images.githubusercontent.com/2631639/115453181-2fadf380-a1d4-11eb-88fb-15fe2dddfef4.png)

Finally I have 4 servers as can be seen here
![image](https://user-images.githubusercontent.com/2631639/115452985-f7a6b080-a1d3-11eb-920e-a6004535f7e7.png)

And expect when lots of jobs are there to have 12 items processing but it does not happen to be like that.
![image](https://user-images.githubusercontent.com/2631639/115458732-f3ca5c80-a1da-11eb-8f27-409deff2b626.png)

Also this does not make sense to me (last 1 hour timeframe)
![image](https://user-images.githubusercontent.com/2631639/115455883-62a5b680-a1d7-11eb-9f16-b236ade4b249.png)
I don't have that many, so why reporting such high numbers?
The following is the current state for the past 1 day
![image](https://user-images.githubusercontent.com/2631639/115455979-7bae6780-a1d7-11eb-9850-82c1263c92aa.png)

**Comment by sseyalioglu** [comment]
I can also see this log "Execution ExpirationManager recovered from the Faulted state after 01:00:05.0250818 and is in the Running state now"
Such a huge time for faulted state. I just wonder what might be holding it up and what it decided to recover after 1 hour. Sorry bombarded you with lots of questions.

**Comment by odinserj** [comment]
Hangfire persists everything to the underlying job storage, and if it doesn't work, nothing will work. Everything matters, even client drivers, especially when talking about corner cases like failovers. For example, a lot of problems in Hangfire.Pro.Redis were related to the underlying client StackExchange.Redis which sometimes didn't reconnect properly, sometimes failed to do anything on thread pool contention, etc.

**Comment by sseyalioglu** [comment]
I agree, and I find out the underlying problem after digging so much.
For some reason, one of the records in Job table is not updatable (don't ask me how), I was unable to update or delete it even after DB restart. I finally decided to dump to file and restore complete table back after truncating and removing FKs. 

Super annoying.

But appreciated your insights.

Another note: With the new DB instance which was Aurora MySQL 2.09.2, hangfire mysql storage does not seem to play ball well where it was perfect with 5.6.mysql_aurora.1.19.5.
I made another DB which is MySQL 8 with much lower instance type and works amazing until now.
I guess it is something to note.
